### PR TITLE
improve handling for structures 

### PR DESCRIPTION
**Pull request by capfredf**
1. fix a bug where struct-out is used with an imported type name
2. implement the design intent that the structure name is not allowed to be used
for the type name,when the latter is provided through #:type-name
3. update the documentation

### CODE REVIEW

**Review by samth — APPROVED** [review]
Good once the tests pass.

**Comment by capfredf** [comment]
some updates: to my recollection, every failed package was successfully built with https://github.com/racket/typed-racket/commit/ff61fa09c2f51f7cf9d2d99bbb5ddeb0ac7823e7.  This morning I simplified the code (https://github.com/racket/typed-racket/commit/190ff08e2d3f86e642b3e9e238be5ebaba8e659d), then one package failed again.

File: typed-racket-lib/typed-racket/typecheck/def-binding.rkt [hunk 1]
```diff
@@ -104,19 +105,19 @@
...
+     (define sname-is-constructor? (and (or extra-constr-name (free-identifier=? sname constr)) #t))
      (define type-is-sname? (free-identifier=? tname internal-id))
      ;; Here, we recursively handle all of the identifiers referenced
      ;; in this static struct info.
      (define-values (constr-defn constr-export-defn constr-new-id constr-aliases)
        (cond
          [(not (identifier? constr))
           (values #'(begin) #'(begin) #f null)]
-         [(free-identifier=? constr internal-id)
+         [(and (free-identifier=? internal-id sname) (free-identifier=? constr internal-id))
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/def-binding.rkt:116]: need a comment

File: typed-racket-lib/typed-racket/utils/struct-info.rkt [hunk 2]
```diff
@@ -100,15 +99,15 @@
 ;; struct-info+type-wrapper is used in either of the follow cases:
 ;; 1. when a structure's name can also represent the structure type.
 ;; 2. when a structure's type name is used as the structure name.
-(struct struct-info+type-wrapper struct-info-wrapper ()
-  #:property prop:procedure
-  (lambda (ins stx)
-    (type-name-error stx))
-  #:extra-constructor-name make-struct-info+type-wrapper)
+;; (struct struct-info+type-wrapper struct-info-wrapper ()
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/utils/struct-info.rkt:102]: delete the commented code

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 3]
```diff
@@ -376,25 +380,38 @@
...
 
   (register-type (struct-names-constructor names) constructor-type)
   (for ([b (in-list def-bindings)])
     (register-type (binding-name b) (def-binding-ty b)))
 
-  (cons struct-binding def-bindings))
+  (cons struct-binding (append (if (free-identifier=? type-name
+                                                      struct-name)
+                                   null
+                                   (list (make-def-struct-stx-binding
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:407]: needs a comment

File: typed-racket-test/succeed/struct-constructors.rkt [hunk 4]
```diff
@@ -1,9 +1,10 @@
-#lang typed/racket/base
+#lang typed/racket
 (module a-mod typed/racket/base
   (provide (except-out (all-defined-out) make-avocado))
   (struct apple ())
   (struct orange () #:constructor-name make-orange)
-  (struct kiwi () #:constructor-name kiwi)
+  (struct kiwi () #:type-name Kiwi)
+  (Kiwi)
```
Comment (capfredf) [inline typed-racket-test/succeed/struct-constructors.rkt:7]: make a super-kiwi using Kiwi

**Pull request merged**